compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/FirSourceChild.kt

```diff
@@ -36,5 +37,48 @@ private fun FirPsiSourceElement<*>.getChild(types: Set<IElementType>, index: Int
 
 private fun FirLightSourceElement.getChild(types: Set<IElementType>, index: Int, depth: Int): FirSourceElement? {
     val visitor = LighterTreeElementFinderByType(treeStructure, types, index, depth)
-    return visitor.find(lighterASTNode)?.toFirLightSourceElement(treeStructure)
+    return visitor.find(lighterASTNode)?.let { withNode(it) }
+}
+
+fun FirSourceElement.getLastChild(type: IElementType, index: Int = 0, depth: Int = -1): FirSourceElement? {
+    return getLastChild(setOf(type), index, depth)
+}
+
+fun FirSourceElement.getLastChild(types: TokenSet, index: Int = 0, depth: Int = -1): FirSourceElement? {
+    return getLastChild(types.types.toSet(), index, depth)
+}
+
+fun FirSourceElement.getLastChild(types: Set<IElementType>, index: Int = 0, depth: Int = -1): FirSourceElement? {
+    return when (this) {
+        is FirPsiSourceElement<*> -> {
+            getLastChild(types, index, depth)
+        }
+        is FirLightSourceElement -> {
+            getLastChild(types, index, depth)
+        }
+        else -> null
+    }
+}
+
+private fun FirPsiSourceElement<*>.getLastChild(types: Set<IElementType>, index: Int, depth: Int): FirSourceElement? {
+    val visitor = PsiElementFinderByType(types, index, depth)
+    return visitor.findLast(psi)?.toFirPsiSourceElement()
+}
+
+private fun FirLightSourceElement.getLastChild(types: Set<IElementType>, index: Int, depth: Int): FirSourceElement? {
+    val visitor = LighterTreeElementFinderByType(treeStructure, types, index, depth)
+    return visitor.findLast(lighterASTNode)?.let { withNode(it) }
+}
+
+private fun FirLightSourceElement.withNode(newNode: LighterASTNode): FirLightSourceElement {
```

Comment:
All this children manipulation code should be removed from PR. BTW in https://github.com/JetBrains/kotlin/pull/4199 the similar code does not influence anything, so I'll remove it from PR before merging it.

Reply:
Removed, but I wonder if `UNSUPPORTED` still needs this. So, similar to wrong modifier, source of `UNSUPPORTED` depends on error kinds, and thus we can't set positioning strategy for that. Instead, we may need to find the specific source, e.g., last child for callable reference expression. I left a TODO there.